### PR TITLE
Teleport action

### PR DESCRIPTION
**Pull request by ThatIsAUsername**
Accidentally started refactoring. Ended up with this. New PR for new discussions.

### CODE REVIEW

File: src/CommandingOfficers/Tech.java [hunk 1]
```diff
@@ -223,23 +225,41 @@ public GameEventQueue getEvents(MapMaster gameMap)
...
+      // Figure out if this caused the enemy's defeat. The TeleportActions can't track
+      // this because multiple units may be smashed at once.
+      Map<Commander, Integer> smashes = new HashMap<Commander, Integer>();
+      for( XYCoord xyc : dropLocs )
+      {
+        Unit r = gameMap.getLocation(xyc).getResident();
+        if( null != r )
+        {
+          Integer oldVal = smashes.putIfAbsent(r.CO, 1);
+          smashes.put(r.CO, (oldVal == null) ? 1 : (oldVal+1));
```
Comment (ThislsAUsername) [inline src/CommandingOfficers/Tech.java:250]: It should be more efficient to just do a normal if statement, here...

File: src/CommandingOfficers/Tech.java [hunk 2]
```diff
@@ -223,23 +225,41 @@ public GameEventQueue getEvents(MapMaster gameMap)
...
+        }
+      }
+      for( Commander co : smashes.keySet() )
+      {
+        if( co.units.size() <= smashes.get(co) )
+          abilityEvents.add(new CommanderDefeatEvent(co));
```
Comment (ThislsAUsername) [inline src/CommandingOfficers/Tech.java:256]: So if you smush his last existing unit with a power, you get two defeat events?

**Comment by ThislsAUsername** [comment]
Overall, I think I prefer the idea of using a mass damage event to notify for power gain, so as to avoid overloading the unit death event unnecessarily.

**Comment by ThatIsAUsername** [comment]
> mass damage event

Certainly doable. The only downside is that we'd have to send `MassDamageEvent`+`UnitDieEvent` to get both death and AP effects.

File: src/CommandingOfficers/Tech.java (hunk 2 above)
Comment (ThatIsAUsername) [reply src/CommandingOfficers/Tech.java:256]: Yes. Realized that after I pushed it up. Not that two defeat events should break anything, but I should probs fix that.

File: src/Engine/GameAction.java [hunk 3]
```diff
@@ -268,7 +276,12 @@ public GameEventQueue getEvents(MapMaster gameMap)
       // If our guy can't survive there, end him.
       if( !unit.model.propulsion.canTraverse(gameMap.getEnvironment(unitDestination)) )
       {
+        ArrayList<Unit> ary = new ArrayList<Unit>();
+        ary.add(unit);
+        boolean fatal = false;
```
Comment (ThislsAUsername) [inline src/Engine/GameAction.java:280]: Needs to be true, or you will get 1 less HP's worth of charge than you should.

File: src/CommandingOfficers/Tech.java [hunk 4]
```diff
@@ -246,13 +246,16 @@ public GameEventQueue getEvents(MapMaster gameMap)
         Unit r = gameMap.getLocation(xyc).getResident();
         if( null != r )
         {
-          Integer oldVal = smashes.putIfAbsent(r.CO, 1);
-          smashes.put(r.CO, (oldVal == null) ? 1 : (oldVal+1));
+          Integer oldVal = smashes.get(r.CO);
+          if( null == oldVal )
+            smashes.put(r.CO, 1);
+          else
+            smashes.put(r.CO, oldVal+1);
```
Comment (ThislsAUsername) [inline src/CommandingOfficers/Tech.java:251]:
```suggestion
          Integer oldVal = smashes.putIfAbsent(r.CO, 1);
          if( null != oldVal ) smashes.put(r.CO, oldVal+1);
```

File: src/CommandingOfficers/Tech.java [hunk 5]
```diff
@@ -223,23 +223,44 @@ public GameEventQueue getEvents(MapMaster gameMap)
...
+      for( XYCoord xyc : dropLocs )
+      {
+        Unit r = gameMap.getLocation(xyc).getResident();
+        if( null != r )
+        {
+          Integer oldVal = smashes.get(r.CO);
+          if( null == oldVal )
+            smashes.put(r.CO, 1);
+          else
+            smashes.put(r.CO, oldVal+1);
```
Comment (ThislsAUsername) [inline src/CommandingOfficers/Tech.java:251]:
```suggestion
          Integer oldVal = smashes.putIfAbsent(r.CO, 1);
          if( null != oldVal ) smashes.put(r.CO, oldVal+1);
```
Ayy, this one worked, I think

File: src/Engine/GameAction.java (hunk 3 above)
Comment (ThatIsAUsername) [reply src/Engine/GameAction.java:280]: Gah. I had it that way, and changed it for testing, and forgot to change it back. Good catch.

**Comment by ThislsAUsername** [comment]
Looks good to me.

**Pull request merged**